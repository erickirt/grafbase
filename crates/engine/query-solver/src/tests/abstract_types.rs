use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
    AGENCY @join__graph(name: "agency", url: "http://localhost:4200/abstract-types/agency")
    BOOKS @join__graph(name: "books", url: "http://localhost:4200/abstract-types/books")
    INVENTORY @join__graph(name: "inventory", url: "http://localhost:4200/abstract-types/inventory")
    MAGAZINES @join__graph(name: "magazines", url: "http://localhost:4200/abstract-types/magazines")
    PRODUCTS @join__graph(name: "products", url: "http://localhost:4200/abstract-types/products")
    REVIEWS @join__graph(name: "reviews", url: "http://localhost:4200/abstract-types/reviews")
    USERS @join__graph(name: "users", url: "http://localhost:4200/abstract-types/users")
}

type Agency
    @join__type(graph: AGENCY, key: "id")
    @join__type(graph: PRODUCTS)
{
    id: ID!
    companyName: String @join__field(graph: AGENCY)
}

type Book implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: BOOKS, key: "id")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    title: String @join__field(graph: BOOKS)
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Book] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type DeliveryEstimates
    @join__type(graph: INVENTORY)
{
    estimatedDelivery: String
    fastestDelivery: String
}

type Group
    @join__type(graph: AGENCY, key: "id")
{
    id: ID!
    name: String
}

type Magazine implements Product & Similar
    @join__implements(graph: INVENTORY, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Product")
    @join__implements(graph: PRODUCTS, interface: "Similar")
    @join__implements(graph: REVIEWS, interface: "Product")
    @join__implements(graph: REVIEWS, interface: "Similar")
    @join__type(graph: INVENTORY, key: "id")
    @join__type(graph: MAGAZINES, key: "id")
    @join__type(graph: PRODUCTS, key: "id")
    @join__type(graph: REVIEWS, key: "id")
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY, requires: "dimensions { size weight }")
    title: String @join__field(graph: MAGAZINES)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    similar: [Magazine] @join__field(graph: PRODUCTS)
    hidden: Boolean @join__field(graph: PRODUCTS)
    publisherType: PublisherType @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
    reviewsOfSimilar: [Review!]! @join__field(graph: REVIEWS, requires: "similar { id }")
}

type ProductDimension
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
{
    size: String
    weight: Float
}

type Query
    @join__type(graph: AGENCY)
    @join__type(graph: BOOKS)
    @join__type(graph: INVENTORY)
    @join__type(graph: MAGAZINES)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
    @join__type(graph: USERS)
{
    books: [Book] @join__field(graph: BOOKS)
    magazines: [Magazine] @join__field(graph: MAGAZINES)
    products: [Product] @join__field(graph: PRODUCTS)
    similar(id: ID!): [Product] @join__field(graph: PRODUCTS)
    review(id: Int!): Review @join__field(graph: REVIEWS)
}

type Review
    @join__type(graph: REVIEWS)
{
    id: Int!
    body: String!
    product: Product
}

type Self
    @join__type(graph: PRODUCTS)
{
    email: String
}

type User
    @join__type(graph: PRODUCTS, key: "email")
    @join__type(graph: USERS, key: "email")
{
    email: ID!
    totalProductsCreated: Int
    name: String @join__field(graph: USERS)
}

interface Product
    @join__type(graph: INVENTORY)
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    id: ID!
    dimensions: ProductDimension @join__field(graph: INVENTORY) @join__field(graph: PRODUCTS)
    delivery(zip: String): DeliveryEstimates @join__field(graph: INVENTORY)
    sku: String @join__field(graph: PRODUCTS)
    createdBy: User @join__field(graph: PRODUCTS)
    hidden: Boolean @inaccessible @join__field(graph: PRODUCTS)
    reviewsCount: Int! @join__field(graph: REVIEWS)
    reviewsScore: Float! @join__field(graph: REVIEWS)
    reviews: [Review!]! @join__field(graph: REVIEWS)
}

interface Similar
    @join__type(graph: PRODUCTS)
    @join__type(graph: REVIEWS)
{
    similar: [Product]
}

union PublisherType
    @join__type(graph: AGENCY)
    @join__type(graph: PRODUCTS)
    @join__unionMember(graph: AGENCY, member: "Agency")
    @join__unionMember(graph: PRODUCTS, member: "Agency")
    @join__unionMember(graph: AGENCY, member: "Group")
    @join__unionMember(graph: PRODUCTS, member: "Self")
 = Agency | Group | Self
"#;

#[test]
fn interface_field_provided_by_implementors() {
    // delivery is not accessible on the Similar interface, but it is on all of its implementors.
    assert_solving_snapshots!(
        "interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        query {
          similar(id: "p1") {
            id
            sku
            delivery(zip: "1234") {
              fastestDelivery
              estimatedDelivery
            }
          }
        }
        "#
    );
}

#[test]
fn nested_interface_field_provided_by_implementors() {
    assert_solving_snapshots!(
        "nested_interface_field_provided_by_implementors",
        SCHEMA,
        r#"
        {
          products {
            id
            reviews {
              product {
                sku
                ... on Magazine {
                  title
                }
                ... on Book {
                  reviewsCount
                }
              }
            }
          }
        }
        "#
    );
}

#[test]
fn unreachable_object() {
    // Group is never reachable.
    assert_solving_snapshots!(
        "unreachable_object",
        SCHEMA,
        r#"
        query {
          products {
            id
            ... on Magazine {
              publisherType {
                ...Publisher
              }
            }
            ... on Book {
              publisherType {
                ...Publisher
              }
            }
          }
        }

        fragment Publisher on PublisherType {
          ... on Agency {
            id
            companyName
          }
          ... on Self {
            email
          }
          ... on Group {
            name
          }
        }
        "#
    );
}
