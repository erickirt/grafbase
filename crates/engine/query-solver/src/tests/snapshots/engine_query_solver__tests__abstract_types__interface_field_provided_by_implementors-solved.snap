---
source: crates/engine/query-solver/src/tests/abstract_types.rs
expression: "digraph {\n    0 [ label = \"root\", color=forestgreen ]\n    1 [ label = \"Product.id\", color=forestgreen ]\n    2 [ label = \"Product.sku\", color=forestgreen ]\n    3 [ label = \"Root#products\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    4 [ label = \"similar#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    5 [ label = \"id#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    6 [ label = \"sku#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    7 [ label = \"DeliveryEstimates.estimatedDelivery\", color=forestgreen ]\n    8 [ label = \"DeliveryEstimates.fastestDelivery\", color=forestgreen ]\n    9 [ label = \"FedEntity#inventory\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    10 [ label = \"delivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    11 [ label = \"estimatedDelivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    12 [ label = \"fastestDelivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    13 [ label = \"dimensions#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    14 [ label = \"*ProductDimension.size\", color=forestgreen ]\n    15 [ label = \"*ProductDimension.weight\", color=forestgreen ]\n    16 [ label = \"weight#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    17 [ label = \"size#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    18 [ label = \"*Book.id\", color=forestgreen ]\n    19 [ label = \"id#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    20 [ label = \"DeliveryEstimates.estimatedDelivery\", color=forestgreen ]\n    21 [ label = \"DeliveryEstimates.fastestDelivery\", color=forestgreen ]\n    22 [ label = \"FedEntity#inventory\", shape=parallelogram, color=dodgerblue, color=forestgreen ]\n    23 [ label = \"delivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    24 [ label = \"estimatedDelivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    25 [ label = \"fastestDelivery#inventory\", shape=box, color=dodgerblue, color=forestgreen ]\n    26 [ label = \"dimensions#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    27 [ label = \"*ProductDimension.size\", color=forestgreen ]\n    28 [ label = \"*ProductDimension.weight\", color=forestgreen ]\n    29 [ label = \"weight#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    30 [ label = \"size#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    31 [ label = \"*Magazine.id\", color=forestgreen ]\n    32 [ label = \"id#products\", shape=box, color=dodgerblue, color=forestgreen ]\n    33 [ label=\"\", style=dashed]\n    0 -> 3 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    3 -> 4 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 5 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    5 -> 1 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 6 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    6 -> 2 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 9 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    9 -> 10 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    10 -> 11 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    11 -> 7 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    10 -> 12 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    12 -> 8 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 13 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    13 -> 16 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    16 -> 15 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    13 -> 17 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    17 -> 14 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 19 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    19 -> 18 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 22 [ label = <<b>1</b>>, color=forestgreen,fontcolor=forestgreen ]\n    22 -> 23 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    23 -> 24 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    24 -> 20 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    23 -> 25 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    25 -> 21 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 26 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    26 -> 29 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    29 -> 28 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    26 -> 30 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    30 -> 27 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    4 -> 32 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    32 -> 31 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n    33 -> 0 [ label = \"\", color=forestgreen,fontcolor=forestgreen ]\n}\n"
---
digraph {
    0 [ label = "root", steiner=1 ]
    1 [ label = "Product.id", steiner=1 ]
    2 [ label = "Product.sku", steiner=1 ]
    3 [ label = "Root#products", steiner=1 ]
    4 [ label = "similar#products", steiner=1 ]
    5 [ label = "id#products", steiner=1 ]
    6 [ label = "sku#products", steiner=1 ]
    7 [ label = "DeliveryEstimates.estimatedDelivery", steiner=1 ]
    8 [ label = "DeliveryEstimates.fastestDelivery", steiner=1 ]
    9 [ label = "FedEntity#inventory", steiner=1 ]
    10 [ label = "delivery#inventory", steiner=1 ]
    11 [ label = "estimatedDelivery#inventory", steiner=1 ]
    12 [ label = "fastestDelivery#inventory", steiner=1 ]
    13 [ label = "dimensions#products", steiner=1 ]
    14 [ label = "*ProductDimension.size", steiner=1 ]
    15 [ label = "*ProductDimension.weight", steiner=1 ]
    16 [ label = "weight#products", steiner=1 ]
    17 [ label = "size#products", steiner=1 ]
    18 [ label = "*Book.id", steiner=1 ]
    19 [ label = "id#products", steiner=1 ]
    20 [ label = "DeliveryEstimates.estimatedDelivery", steiner=1 ]
    21 [ label = "DeliveryEstimates.fastestDelivery", steiner=1 ]
    22 [ label = "FedEntity#inventory", steiner=1 ]
    23 [ label = "delivery#inventory", steiner=1 ]
    24 [ label = "estimatedDelivery#inventory", steiner=1 ]
    25 [ label = "fastestDelivery#inventory", steiner=1 ]
    26 [ label = "dimensions#products", steiner=1 ]
    27 [ label = "*ProductDimension.size", steiner=1 ]
    28 [ label = "*ProductDimension.weight", steiner=1 ]
    29 [ label = "weight#products", steiner=1 ]
    30 [ label = "size#products", steiner=1 ]
    31 [ label = "*Magazine.id", steiner=1 ]
    32 [ label = "id#products", steiner=1 ]
    33 [ label="", style=dashed]
    0 -> 3 [ cost=1, steiner=1]
    3 -> 4 [ cost=0, steiner=1]
    4 -> 5 [ cost=0, steiner=1]
    5 -> 1 [ cost=0, steiner=1]
    4 -> 6 [ cost=0, steiner=1]
    6 -> 2 [ cost=0, steiner=1]
    4 -> 9 [ cost=1, steiner=1]
    9 -> 10 [ cost=0, steiner=1]
    10 -> 11 [ cost=0, steiner=1]
    11 -> 7 [ cost=0, steiner=1]
    10 -> 12 [ cost=0, steiner=1]
    12 -> 8 [ cost=0, steiner=1]
    4 -> 13 [ cost=0, steiner=1]
    13 -> 16 [ cost=0, steiner=1]
    16 -> 15 [ cost=0, steiner=1]
    13 -> 17 [ cost=0, steiner=1]
    17 -> 14 [ cost=0, steiner=1]
    4 -> 19 [ cost=0, steiner=1]
    19 -> 18 [ cost=0, steiner=1]
    4 -> 22 [ cost=1, steiner=1]
    22 -> 23 [ cost=0, steiner=1]
    23 -> 24 [ cost=0, steiner=1]
    24 -> 20 [ cost=0, steiner=1]
    23 -> 25 [ cost=0, steiner=1]
    25 -> 21 [ cost=0, steiner=1]
    4 -> 26 [ cost=0, steiner=1]
    26 -> 29 [ cost=0, steiner=1]
    29 -> 28 [ cost=0, steiner=1]
    26 -> 30 [ cost=0, steiner=1]
    30 -> 27 [ cost=0, steiner=1]
    4 -> 32 [ cost=0, steiner=1]
    32 -> 31 [ cost=0, steiner=1]
    33 -> 0 [ cost=0, steiner=1]
}
